File: skills/edit/scripts/edit_records.py
"""Edit or delete records in the estimator history CSV.

Reuses build_csv's parser/HEADERS/writer. Driven by an instructions file:
  {"edits": [{"key": "ABC-1", "fields": {"summary": "...", ...}}], "deletes": ["ABC-2"]}
Prints JSON: {"updated":[...],"deleted":[...],"missing":[...],"row_count":N}
`key` is read-only; unknown field names are rejected (validated before any change).
"""

import json
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "scripts"))
import build_csv  # noqa: E402

FIELD_KEYS = build_csv.FIELD_KEYS
COL = {fk: i for i, fk in enumerate(FIELD_KEYS)}


def _dedup(seq):
    seen = set()
    out = []
    for x in seq:
        if x not in seen:
            seen.add(x)
            out.append(x)
    return out


def apply_changes(rows, edits, deletes):
    # Validate everything before mutating, so a bad instruction changes nothing.
    for edit in edits:
        for fk in (edit.get("fields") or {}):
            if fk == "key":
                raise ValueError("'key' is read-only and cannot be edited")
            if fk not in COL:
                raise ValueError("unknown field: %r" % fk)
    edit_keys = {str(e.get("key", "")).strip().upper() for e in edits}
    delete_keys_set = {str(k).strip().upper() for k in deletes}
    overlap = (edit_keys & delete_keys_set) - {""}
    if overlap:
        raise ValueError("key(s) appear in both edits and deletes: %s" % ", ".join(sorted(overlap)))
    index = {}
    for row in rows:
        if row:
            index[row[0]] = row
    updated = []
    delete_keys = []
    missing = []
    for edit in edits:
        k = str(edit.get("key", "")).strip().upper()
        if not k:
            continue
        if k in index:
            row = index[k]
            for fk, val in (edit.get("fields") or {}).items():
                row[COL[fk]] = "" if val is None else str(val)
            updated.append(k)
        else:
            missing.append(k)
    for key in deletes:
        k = str(key).strip().upper()
        if not k:
            continue
        if k in index:
            delete_keys.append(k)
        else:
            missing.append(k)
    delete_set = set(delete_keys)
    result_rows = [row for row in rows if row and row[0] not in delete_set]
    return {
        "rows": result_rows,
        "updated": _dedup(updated),
        "deleted": _dedup(delete_keys),
        "missing": _dedup(missing),
    }


def main(argv):
    # argv is sys.argv: argv[0] is the script name, argv[1]/argv[2] are the args
    if len(argv) != 3:
        sys.exit("Usage: edit_records.py <history.csv> <instructions.json>")
    history_csv, instructions_path = argv[1], argv[2]
    with open(instructions_path, encoding="utf-8") as f:
        instructions = json.load(f)
    rows = build_csv.read_existing(history_csv)
    try:
        result = apply_changes(rows, instructions.get("edits") or [], instructions.get("deletes") or [])
    except ValueError as e:
        sys.exit("error: %s" % e)
    build_csv.write_csv(history_csv, result["rows"])
    summary = {
        "updated": result["updated"],
        "deleted": result["deleted"],
        "missing": result["missing"],
        "row_count": len(result["rows"]),
    }
    sys.stdout.buffer.write(
        (json.dumps(summary, ensure_ascii=False, separators=(",", ":")) + "\n").encode("utf-8")
    )


if __name__ == "__main__":
    main(sys.argv)

File: skills/edit/scripts/edit_records.mjs
// Edit or delete records in the estimator history CSV.
// Reuses build_csv's parser/HEADERS/writer. Instructions:
//   {"edits":[{"key":"ABC-1","fields":{"summary":"..."}}],"deletes":["ABC-2"]}
// Prints {"updated":[...],"deleted":[...],"missing":[...],"row_count":N}.
import { HEADERS, FIELD_KEYS as _FIELD_KEYS, readExisting, writeCsv } from "../../../scripts/build_csv.mjs";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

export const FIELD_KEYS = _FIELD_KEYS;
const COL = new Map(FIELD_KEYS.map((fk, i) => [fk, i]));

function dedup(seq) {
  const seen = new Set();
  const out = [];
  for (const x of seq) if (!seen.has(x)) { seen.add(x); out.push(x); }
  return out;
}

export function applyChanges(rows, edits, deletes) {
  // Validate before mutating so a bad instruction changes nothing.
  for (const edit of edits) {
    for (const fk of Object.keys(edit.fields || {})) {
      if (fk === "key") throw new Error("'key' is read-only and cannot be edited");
      if (!COL.has(fk)) throw new Error(`unknown field: ${fk}`);
    }
  }
  const editKeys = new Set(edits.map((e) => String(e.key ?? "").trim().toUpperCase()));
  const deleteKeysSet = new Set(deletes.map((k) => String(k).trim().toUpperCase()));
  const overlap = [...editKeys].filter((k) => k && deleteKeysSet.has(k)).sort();
  if (overlap.length) throw new Error(`key(s) appear in both edits and deletes: ${overlap.join(", ")}`);
  const index = new Map();
  for (const row of rows) if (row && row.length) index.set(row[0], row);
  const updated = [];
  const deleteKeys = [];
  const missing = [];
  for (const edit of edits) {
    const k = String(edit.key ?? "").trim().toUpperCase();
    if (!k) continue;
    if (index.has(k)) {
      const row = index.get(k);
      for (const [fk, val] of Object.entries(edit.fields || {})) {
        row[COL.get(fk)] = val === null || val === undefined ? "" : String(val);
      }
      updated.push(k);
    } else {
      missing.push(k);
    }
  }
  for (const key of deletes) {
    const k = String(key).trim().toUpperCase();
    if (!k) continue;
    if (index.has(k)) deleteKeys.push(k);
    else missing.push(k);
  }
  const deleteSet = new Set(deleteKeys);
  const resultRows = rows.filter((row) => row && row.length && !deleteSet.has(row[0]));
  return { rows: resultRows, updated: dedup(updated), deleted: dedup(deleteKeys), missing: dedup(missing) };
}

export function main(argv) {
  // argv is already sliced (process.argv.slice(2)): [historyCsv, instructionsPath]
  if (argv.length !== 2) {
    console.error("Usage: edit_records.mjs <history.csv> <instructions.json>");
    process.exit(1);
  }
  const [historyCsv, instructionsPath] = argv;
  const instructions = JSON.parse(readFileSync(instructionsPath, "utf8"));
  const rows = readExisting(historyCsv);
  let result;
  try {
    result = applyChanges(rows, instructions.edits || [], instructions.deletes || []);
  } catch (e) {
    console.error(`error: ${e.message}`);
    process.exit(1);
  }
  writeCsv(historyCsv, result.rows);
  const summary = {
    updated: result.updated,
    deleted: result.deleted,
    missing: result.missing,
    row_count: result.rows.length,
  };
  console.log(JSON.stringify(summary));
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2));
}
